reinforcement_learning/rlclientlib/utility/watchdog.h

```diff
@@ -0,0 +1,61 @@
+#pragma once
+#include <thread>
+#include <map>
+#include <mutex>
+
+#include "api_status.h"
+#include "error_callback_fn.h"
+
+#include <atomic>
+#include "interruptable_sleeper.h"
+
+namespace reinforcement_learning {
+  namespace utility {
+
+
+    class watchdog {
+    public:
+      explicit watchdog(error_callback_fn* error_callback = nullptr);
+      ~watchdog();
+
+      void register_thread(std::thread::id const& thread_id, std::string const& thread_name, long long const timeout);
+      void unregister_thread(std::thread::id const thread_id);
+      void check_in(std::thread::id const thread_id);
+
+      int start(api_status* status);
+      void stop();
+      void loop();
+
+      bool should_report_unhandled_background_error() const;
```

Comment:
>should_report_unhandled_background_error [](start = 11, length = 40)

Why ask if the watchdog handles background errors? I suggest just sending the "set_unhandled_background_error" and allowing the watchdog to No-op if they want.

Reply:
This check is for live_model to check if a background error has been reported so that they can then surface that. Perhaps the naming is not so good here, I'll update it to hopefully be a bit clearer